examples/nlp/intent_detection_slot_tagging/joint_intent_slot_infer.py

```diff
@@ -103,29 +104,181 @@ def get_preds(logits):
+            logging.info(str)
+            str = ''
+
+    logging.info('----------------')
+    cnt = 0
+    for i in range(len(intent_preds)):
+        cur_slot_pred = slot_preds[i][subtokens_mask[i]]
+        cur_slot_label = slot_labels[i][subtokens_mask[i]]
+        if not np.all(cur_slot_pred == cur_slot_label):
+            query = queries[i].split('\t')[0]
+            logging.info(
+                f'{query} - intent: {intent_dict[intent_labels[i]]} - predicted: {intent_dict[intent_preds[i]]}'
+            )
+            logging.info(cur_slot_pred)
+            logging.info(cur_slot_label)
+            cnt = cnt + 1
+            if cnt >= limit:
+                break
+
+
+# check non compliance of B- and I- slots,
+def check_problematic_slots(slot_preds_list, slot_dict):
+    slot_dict = open(slot_dict, 'r').readlines()
+    cnt = 0
+
+    # for sentence in slot_preds:
+    # slots = sentence.split(" ")
+    sentence = slot_preds_list
+    for i in range(len(sentence)):
+        slot_name = slot_dict[int(sentence[i])]
+        if slot_name.startswith("I-"):
+            prev_slot_name = slot_dict[int(sentence[i - 1])]
+            if slot_name[2:] != prev_slot_name[2:]:
+                print("Problem: " + slot_name + " - " + prev_slot_name)
+                cnt += 1
+    print("Total problematic slots: " + str(cnt))
+
+
+def errors_per_class(cm, dict):
+    size = cm.shape[0]
+    confused_per_class = {}
+    total_errors = 0
+    for class_num in range(size):
+        sum = 0
+        for i in range(size):
+            if i != class_num:
+                sum += cm[class_num][i]
+                sum += cm[i][class_num]
+        confused_per_class[dict[class_num]] = sum
+        total_errors += sum
+        # logging.info(f'{dict[class_num]} - {sum}')
+
+    logging.info(f'Total errors (multiplied by 2): {total_errors}')
+    sorted_confused_per_class = sorted(confused_per_class.items(), key=lambda x: x[1], reverse=True)
+    for conf_str in sorted_confused_per_class:
+        logging.info(conf_str)
+
+
+def analyze_confusion_matrix(cm, dict, max_pairs=10):
+    threshold = 5
+    confused_pairs = {}
+    size = cm.shape[0]
+    for i in range(size):
+        res = cm[i].argsort()
+        for j in range(size):
+            pos = res[size - j - 1]
+            # no confusion - same row and column
+            if pos == i:
+                continue
+            elif cm[i][pos] >= threshold:
+                str = f'{dict[i]} -> {dict[pos]}'
+                confused_pairs[str] = cm[i][pos]
+            else:
+                break
+
+    # sort my max confusions and print first max_pairs
+    sorted_confused_pairs = sorted(confused_pairs.items(), key=lambda x: x[1], reverse=True)
+    for i, pair_str in enumerate(sorted_confused_pairs):
+        if i >= max_pairs:
+            break
+        logging.info(pair_str)
+
+
+intent_logits, slot_logits, loss_mask, subtokens_mask, intent_labels, slot_labels_unmasked = [
```

Comment:
Move those lines into main() function:
if __name__ == "__main__":

Reply:
All our code in this package is not written in this way. I do not want to make too big refactoring at this stage.